### DISCUSSION TITLE
Window title does not reset after exiting neovim

### DISCUSSION THREAD

**Discussion by strayer** [kovidgoyal/kitty#3636] [Q&A] [answered]
Not sure if this is an issue so I'm asking here first. I'm using https://github.com/knubie/vim-kitty-navigator to use the same hotkeys for navigating neovim and kitty windows. This works fine, but for some reason the window title is not reset when I'm exiting neovim. This works fine in iTerm2, so I assume there is either a bug or I misconfigured something.

I'm setting the window title via neovim like this in the init.vim:

```
set title
let &titlestring='%t - nvim'
```

<details>
<summary>kitty --debug-config</summary>

```
kitty 0.20.3 created by Kovid Goyal
Darwin khitomer 20.4.0 Darwin Kernel Version 20.4.0: Thu Apr 22 21:46:47 PDT 2021; root:xnu-7195.101.2~1/RELEASE_X86_64 x86_64
ProductName:	macOS ProductVersion:	11.3.1 BuildVersion:	20E241
Loaded config files: /Users/strayer/.config/kitty/kitty.conf

Config options different from defaults:
active_tab_background   Color(red=122, green=162, blue=247)
active_tab_foreground   Color(red=31, green=35, blue=53)
background              Color(red=36, green=40, blue=59)
bold_font               Iosevka Fixed Bold
bold_italic_font        Iosevka Fixed Bold Italic
color0                  Color(red=29, green=32, blue=47)
color1                  Color(red=247, green=118, blue=142)
color10                 Color(red=158, green=206, blue=106)
color11                 Color(red=224, green=175, blue=104)
color12                 Color(red=122, green=162, blue=247)
color13                 Color(red=187, green=154, blue=247)
color14                 Color(red=125, green=207, blue=255)
color15                 Color(red=192, green=202, blue=245)
color16                 Color(red=255, green=158, blue=100)
color17                 Color(red=219, green=75, blue=75)
color2                  Color(red=158, green=206, blue=106)
color3                  Color(red=224, green=175, blue=104)
color4                  Color(red=122, green=162, blue=247)
color5                  Color(red=187, green=154, blue=247)
color6                  Color(red=125, green=207, blue=255)
color7                  Color(red=169, green=177, blue=214)
color8                  Color(red=65, green=72, blue=104)
color9                  Color(red=247, green=118, blue=142)
cursor                  Color(red=192, green=202, blue=245)
cursor_shape            2
enabled_layouts         ['splits:split_axis=horizontal']
font_family             Iosevka Fixed
font_size               14.0
foreground              Color(red=192, green=202, blue=245)
inactive_tab_background Color(red=41, green=46, blue=66)
inactive_tab_foreground Color(red=84, green=92, blue=126)
italic_font             Iosevka Fixed Italic
macos_titlebar_color    1
selection_background    Color(red=54, green=74, blue=130)
selection_foreground    Color(red=192, green=202, blue=245)
symbol_map              {(58874, 58923): 'Symbols Nerd Font', (59136, 59333): 'Symbols Nerd Font', (61440, 62176): 'Symbols Nerd Font', (57856, 58025): 'Symbols Nerd Font', (62720, 64838): 'Symbols Nerd Font', (58112, 58347): 'Symbols Nerd Font', (62464, 62632): 'Symbols Nerd Font', (9829, 9829): 'Symbols Nerd Font', (9889, 9889): 'Symbols Nerd Font', (62076, 62076): 'Symbols Nerd Font', (57507, 57507): 'Symbols Nerd Font', (57524, 57544): 'Symbols Nerd Font', (57548, 57554): 'Symbols Nerd Font', (57556, 57556): 'Symbols Nerd Font', (9211, 9214): 'Symbols Nerd Font', (11096, 11096): 'Symbols Nerd Font', (62208, 62227): 'Symbols Nerd Font', (57344, 57357): 'Symbols Nerd Font'}
sync_to_monitor         False
url_color               Color(red=115, green=218, blue=202)
Added shortcuts:
	shift+left KeyAction(func='previous_tab', args=())
	shift+right KeyAction(func='next_tab', args=())
	shift+up KeyAction(func='layout_action', args=['rotate', ()])
	ctrl+h KeyAction(func='kitten', args=['pass_keys.py', 'neighboring_window left   ctrl+h "^.* - nvim$"'])
	ctrl+j KeyAction(func='kitten', args=['pass_keys.py', 'neighboring_window bottom ctrl+j "^.* - nvim$"'])
	ctrl+k KeyAction(func='kitten', args=['pass_keys.py', 'neighboring_window top    ctrl+k "^.* - nvim$"'])
	ctrl+l KeyAction(func='kitten', args=['pass_keys.py', 'neighboring_window right  ctrl+l "^.* - nvim$"'])
	ctrl+left KeyAction(func='move_window', args=['left'])
	ctrl+right KeyAction(func='move_window', args=['right'])
	ctrl+up KeyAction(func='move_window', args=['top'])
	ctrl+down KeyAction(func='move_window', args=['bottom'])
	cmd+d KeyAction(func='launch', args=['--location=vsplit'])
	cmd+tab KeyAction(func='next_tab', args=())
	shift+cmd+tab KeyAction(func='previous_tab', args=())
Changed shortcuts:
	shift+cmd+d KeyAction(func='launch', args=['--location=hsplit'])
```

</details>

While testing this with `kitty --config NONE` I noticed these errors:

```
[137 15:11:05.935658] [PARSE ERROR] CSI code 0x74 has 3 > 2 parameters
[137 15:11:06.418874] WARNING: Your system's OpenGL implementation does not have glCopyImageSubData, falling back to a slower implementation
[137 15:11:09.974108] [PARSE ERROR] CSI code 0x74 has 3 > 2 parameters
```

I also tried setting the title manually with `echo -en "\033]0;title\a"` before starting neovim, but that didn't change anything.

iTerm2:

After starting iTerm:

<img width="564" alt="grafik" src="https://user-images.githubusercontent.com/310624/118493573-8bab5f80-b721-11eb-85c9-38607622655f.png">

After starting neovim:

<img width="564" alt="grafik" src="https://user-images.githubusercontent.com/310624/118493607-9534c780-b721-11eb-88b2-20c72ff402b8.png">

After exiting neovim:

<img width="574" alt="grafik" src="https://user-images.githubusercontent.com/310624/118493639-9d8d0280-b721-11eb-934c-b691cd6fd476.png">

After launching kitty:

<img width="458" alt="grafik" src="https://user-images.githubusercontent.com/310624/118493717-b09fd280-b721-11eb-8bb8-238dec081fbe.png">

After starting neovim:

<img width="454" alt="grafik" src="https://user-images.githubusercontent.com/310624/118493743-b990a400-b721-11eb-95f2-1bcf191b6192.png">

After leaving neovim:

<img width="455" alt="grafik" src="https://user-images.githubusercontent.com/310624/118493767-c01f1b80-b721-11eb-907a-b3ad52fb657b.png">

**Comment by kovidgoyal** [comment]
The terminal knows nothing about when programs launched by shells start or quit. If the title is being reset it means either that the program itself is resetting it or the shell is doing so after the program quits. If I had to guess, I would guess the shell. Presumably you have some iterm specific config in your shell rc files. I suggest  you add the following to .zshrc, which will work aith *any* terminal not just iterm

```sh
    # Show the current working directory in the terminal window titlebar
    function set-title-precmd () { printf "\033]2;${PWD/$HOME/~}\007" }
    function set-title-preexec() { printf "\e]2;%s\a" "$1" }
    autoload -Uz add-zsh-hook
    add-zsh-hook precmd set-title-precmd
    add-zsh-hook preexec set-title-preexec
```

**Reply by strayer** [reply]
FWIW this also works on Terminal.app (though the title there has some more stuff at the end, not sure whats going on there).

Neovim should utilize the title stack (https://github.com/kovidgoyal/kitty/commit/3067103b188c55d594918d1525d397cb523f8fea) for this, I don't think there is anything specific to iTerm or Terminal.app in my zsh configuration.

**Comment by kovidgoyal** [comment, marked as answer]
It works fine for me with vim, I dont use nvim so cannot comment on it. Steps I tried:

1) Run kitty sh to run a basic POSIX sheel that does not manipulate window titles
2) echo -ne '\033]2;moose\007'  to change the window title to moose
3) Start vim
4) window title is changed by vim
5) Quit vim
6) window title is again moose

**Reply by strayer** [reply]
This is weird. I just tested this too with /bin/sh as a shell and can confirm what you are seeing. vim resets the title just fine, neovim doesn't. I started both with `--clean` so they don't use any config file, so this really is related to neovim. I'll try to get to the bottom of this.

Thank you for taking time to look into this!

**Reply by kovidgoyal** [reply]
You can use --dump-commands to have kitty print out all the escape codes it receives. That should tell you is nvim is using the title stack or not.

**Reply by strayer** [reply]
Thanks for the tip!

This is what happens in vim when I do `:set title` and `:q`:

```
screen_set_mode 25 1
draw s
draw e
draw t
draw  
draw t
draw i
draw t
draw l
draw e
screen_carriage_return
screen_reset_mode 25 1
set_title [No Name] - VIM
screen_cursor_position 1 1
screen_set_mode 25 1
screen_reset_mode 25 1
screen_cursor_position 84 1
screen_erase_in_line 0 0
screen_cursor_position 84 1
draw :
screen_set_mode 25 1
draw q
screen_carriage_return
screen_reset_mode 25 1
screen_reset_mode 2004 1
screen_xtmodkeys 4 0
set_title Thanks for flying Vim
screen_manipulate_title_stack 23 2
screen_manipulate_title_stack 23 1
screen_manipulate_title_stack 22 2
screen_manipulate_title_stack 22 1
screen_manipulate_title_stack 23 2
screen_manipulate_title_stack 23 1
screen_cursor_position 84 1
screen_erase_in_line 0 0
screen_cursor_position 84 1
screen_reset_mode 2004 1
screen_reset_mode 1 1
screen_normal_keypad_mode
screen_set_mode 25 1
screen_xtmodkeys 4 0
screen_reset_mode 1049 1
```

This is what happens in neovim:

```
draw :
set_dynamic_color 112 
screen_set_cursor 2 32
screen_set_mode 25 1
screen_reset_mode 25 1
draw s
screen_set_mode 25 1
screen_reset_mode 25 1
draw e
screen_set_mode 25 1
screen_reset_mode 25 1
draw t
screen_set_mode 25 1
screen_reset_mode 25 1
screen_cursor_forward 1
screen_set_mode 25 1
screen_reset_mode 25 1
draw t
screen_set_mode 25 1
screen_reset_mode 25 1
draw i
screen_set_mode 25 1
screen_reset_mode 25 1
draw t
screen_set_mode 25 1
screen_reset_mode 25 1
draw l
screen_set_mode 25 1
screen_reset_mode 25 1
draw e
screen_set_mode 25 1
screen_reset_mode 25 1
screen_carriage_return
screen_cursor_position 84 1
screen_set_mode 25 1
screen_reset_mode 25 1
set_title [Unbenannt] - NVIM
set_dynamic_color 112 
screen_set_cursor 2 32
screen_set_mode 25 1
screen_reset_mode 25 1
screen_cursor_position 1 1
screen_set_mode 25 1
screen_reset_mode 25 1
screen_cursor_position 84 231
draw :
screen_cursor_position 1 1
screen_set_mode 25 1
screen_reset_mode 25 1
screen_cursor_down 83
screen_erase_characters 231
draw :
set_dynamic_color 112 
screen_set_cursor 2 32
screen_set_mode 25 1
screen_reset_mode 25 1
draw q
screen_set_mode 25 1
screen_reset_mode 25 1
screen_carriage_return
screen_cursor_position 84 1
screen_set_mode 25 1
screen_reset_mode 25 1
set_dynamic_color 112 
screen_set_cursor 2 32
screen_set_mode 25 1
screen_reset_mode 25 1
set_dynamic_color 112 
screen_set_cursor 2 32
screen_designate_charset 0 66
select_graphic_rendition 0 
screen_set_mode 25 1
screen_reset_mode 1 1
screen_reset_mode 1049 1
CSI code 0x74 has 3 > 2 parameters
screen_reset_mode 2004 1
screen_reset_mode 1004 1
screen_set_mode 25 1
```

It really looks like neovim does not send the CSI codes to reset the title. Weird!

**Reply by kovidgoyal** [reply]
That log shows the problem: CSI code 0x74 has 3 > 2 parameters  nvim is sending the escap ecode with an extra parameter 3 instead of 2. vim uses 2. kitty currently handles only 2.

**Reply by kovidgoyal** [reply]
And as can be seen here: https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h3-Functions-using-CSI-_-ordered-by-the-final-character_s_   (search for Ps = 2 2 ; 0) the title stach CSI code should have 2 not 3 parameters. You should report the bug to nvim, though I can probably also modify kitty to ignore the third parameter. However I would rather that programs send the correct escape codes.

**Reply by strayer** [reply]
I already reported the bug earlier (https://github.com/neovim/neovim/issues/14573), but added your findings there too. Hopefully the neovim team can help resolve this.

Thanks a lot for investing time to look into this!

**Reply by jamessan** [reply]
The window manipulation CSI has 3 parameters:

```
CSI Ps ; Ps ; Ps t
          Window manipulation (XTWINOPS)
```

Just because that specific win op doesn't use all 3 parameters, doesn't mean it's buggy for nvim to send the implied 0 value for the 3rd parameter.

**Reply by kovidgoyal** [reply]
No XTWINOPS has a variable number of parameters, between 1 and 3. The number of parameters is determined by the value of the first parameter. Someday in the future the escape code might be extended to actually use the third parameter when setting titles. And there is no implied zero in the definition. In fact there are explicit zeros for the second parameter.

**Reply by jamessan** [reply]
XTWINOPS has 3 parameters.  Yes, some of the parameters may not be used yet for some of the sub-commands.  However, I can send `\e[22t` and that implicitly means `\e22;0t`.

By that same logic, if it does get a new parameter, then either everything not sending the third parameter becomes insta-buggy or the 0 value is the same behavior as today and anything sending `\e[22t` or `\e[22;0t` get the implied 0 value.  So, I don't see why sending the explicit third parameter is a problem.

Also kitty 0.19.3 behaves as expected with your moose example, so it seems kitty has become more strict about this in the interim.

**Comment by kovidgoyal** [comment]
On Tue, May 18, 2021 at 05:02:16AM -0700, James McCoy wrote:
> XTWINOPS has 3 parameters.  Yes, some of the parameters may not be used yet for some of the sub-commands.  However, I can send `\e[22t` and that implicitly means `\e22;0t`.

Says, who? I can see nowhere in the specification that says that. It
might be an *implied* behavior but it certainly isn't *specified*.

> 
> By that same logic, if it does get a new parameter, then either everything not sending the third parameter becomes insta-buggy or the 0 value is the same behavior as today and anything sending `\e[22t` or `\e[22;0t` get the implied 0 value.  So, I don't see why sending the explicit third parameter is a problem.

Umm why are you arguing about this? Is it your position that nvim *should*
be sending an extra meaningless two bytes? If so, why? And if not, why
not spend this energy sending a patch for nvim instead?

> 
> Also kitty 0.19.3 behaves as expected with your moose example, so it seems kitty has become more strict about this in the interim.

Indeed, I have been gradually tightening kitty's parser with time.

**Reply by jamessan** [reply]
> Umm why are you arguing about this? Is it your position that nvim *should* be sending an extra meaningless two bytes?

No, I'm arguing that your parser is being unnecessarily strict.

> And if not, why not spend this energy sending a patch for nvim instead?

I've already done that.

**Comment by kovidgoyal** [comment]
On Tue, May 18, 2021 at 05:55:23AM -0700, James McCoy wrote:
> > Umm why are you arguing about this? Is it your position that nvim *should* be sending an extra meaningless two bytes?
> 
> No, I'm arguing that your parser is being unnecessarily strict.

The terminal ecosystem could really do with a little strictness.

> 
> > And if not, why not spend this energy sending a patch for nvim instead?
> 
> I've already done that.

Excellent, thanks, then this issue is resolved. If someday there is a
plague of software sending meaningless zeros with this escape code, I
will reconsider relaxing the parser.

**Reply by strayer** [reply]
Thank you @kovidgoyal and @jamessan for taking the time to fix this! I appreciate it a lot!
